### PR TITLE
Release saved variable from DifferentiableGraphBackward

### PR DESCRIPTION
**Pull request by sujoysaraswati**
When the backward ops execute via the autograd engine evaluate_function(), the fn.release_variables() is called to release the SavedVariables. For the eager mode ops, this releases the saved inputs that was required for backward grad function. However, with TorchScript, we get a DifferentableGraph and the DifferentiableGraphBackward() doesn't implement a release_variables(). This leads to the SavedVariables to be alive longer. Implement release_variables() for DifferentiableGraphBackward to release these SavedVariables  early.

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
From an autograd point of view, this looks good to me.
Note that this is most likely missing code that handle the case where these were freed but the user tried to backward through the graph again. The regular autograd throws an error saying that you should run the first backward with retain_graph=True.
You can test this by doing a single forward and two backwards in a row on the same graph without setting retain_graph=True.
Could you add that test to make sure it works fine?


Also since this happens only at the end of the Node, and that Node can contain the backward for a significant part of the graph, the memory usage might still be higher than eager autograd. But fixing that would be much more complicated and out of the scope of this PR I think.

**Comment by SujoysGithub** [comment]
> From an autograd point of view, this looks good to me.
> Note that this is most likely missing code that handle the case where these were freed but the user tried to backward through the graph again. The regular autograd throws an error saying that you should run the first backward with retain_graph=True.
> You can test this by doing a single forward and two backwards in a row on the same graph without setting retain_graph=True.
> Could you add that test to make sure it works fine?

I tested with the tests test_autograd.py::TestAutograd::test_backward_twice_with_saved_values and test_backward_twice_without_saved_values. They do two backwards in a row without retain_graph=True. They are passing - would they be sufficient for the tests you described above?

> Also since this happens only at the end of the Node, and that Node can contain the backward for a significant part of the graph, the memory usage might still be higher than eager autograd. But fixing that would be much more complicated and out of the scope of this PR I think.

Agree, the memory for the saved variables are freed only at the end of the backward op, it would be great if we can free the memory even earlier, but I haven't explored it.

**Comment by albanD** [comment]
> They are passing - would they be sufficient for the tests you described above?

Yes something like that for an symbolically diffed code sample would work.
Out of curiosity, does it raise the exact same error as the non-scripted version? Where does this check happens?

**Comment by SujoysGithub** [comment]
> > They are passing - would they be sufficient for the tests you described above?
> 
> Yes something like that for an symbolically diffed code sample would work.

I have added a test in test_jit.py in this pull request. It is expected to raise the runtime error with second backward() when retain_graph is false, and should pass through when it is true.

> Out of curiosity, does it raise the exact same error as the non-scripted version? Where does this check happens?

It comes from SavedVariable::unpack, I get the exception from this place.

**Review by albanD — APPROVED** [review]
Ho right it properly unpacks stuff in the Node.

LGTM!

**Comment by SujoysGithub** [comment]
Let me know if I need to do anything else for this patch - is the next step to get more reviews?

**Comment by albanD** [comment]
We should get a review from someone in the jit team.

**Review by zdevito — APPROVED** [review]
Looks good to me. I don't see any obvious ways this would behave differently in JIT than it does in eager.

**Comment by albanD** [comment]
@SujoysGithub could you rebase on top of master to have the CI running again please? I see some test failures after rebasing.

File: test/test_jit.py [hunk 1]
```diff
@@ -950,6 +950,38 @@ def test_torch_autograd_backward_with_grad_tensors(input):
         checkBackwardScript(test_torch_autograd_backward, (inp,))
         checkBackwardScript(test_torch_autograd_backward_with_grad_tensors, (inp,))
 
+    def test_script_backward_twice(self):
+        def checkBackwardTwiceScript(fn, inputs, retain_graph_=False):
+            torch._C._jit_set_profiling_executor(False)
```
Comment (gmagogsfm) [inline test/test_jit.py:955]: OOC, why do you choose to set the executor state here without restoring it to previous value?

I noticed that this test interferes with one of tests that I am trying to add. Would it be better if we only run this test when Profiling Executor is turned off, using @unittest.skipIf?

Comment (SujoysGithub) [reply test/test_jit.py:955]: I can skip it when profiling executor is turned on. Are the tests run with profiling executor off under some testsuite already?

Comment (gmagogsfm) [reply test/test_jit.py:955]: In that case, I think you need to decorate your test case with

@unittest.skipIf(GRAPH_EXECUTOR == ProfilingMode.LEGACY, "REASON REASON REASON")

Comment (SujoysGithub) [reply test/test_jit.py:955]: Raised https://github.com/pytorch/pytorch/pull/45893

**Pull request closed without merging**